public/src/components/ToastManager.jsx

```diff
@@ -0,0 +1,198 @@
+    `,
+    error: `
+      background-color: #EF6F6C;
+    `,
+  };
+  return flavors[flavor];
+};
+
+export const ToastContext = React.createContext();
+const ToastRack = styled.div`
+  z-index: 1;
+  position: fixed;
+  right: 0;
+  width: 250px;
+  bottom: 0;
+  padding: 1em;
+`;
+
+const StyledToast = styled.div`
+  @keyframes styledtoast {
+    0% {
+      transform: translateY(20px)
+    }
+    100% {
+      transform: translateY(0px)
+    }
+  }
+  @keyframes cancelled {
+  0% {
+    transform: translateX(0px);
+  }
+  90% {
+    transform: translateX(0px);
+  }
+  100% {
+    transform: translateX(300px);
+  }
+}
+  @keyframes immediatecancel {
+  0% {
+    transform: translateX(0px);
+  }
+  100% {
+    transform: translateX(300px);
+  }
+}
+
+  display: flex;
+  border: 1px solid #54457F;
+  width: 250px;
+  cursor: pointer;
+  animation: styledtoast .3s ease-out, cancelled 4.5s ease-out .3s;
+  animation-fill-mode: forwards;  
+  border-radius: .25em;
+  background-color: white;
+  box-shadow: 0px 4px 3px 3px rgba(0,0,0, .1);
+  & + & {
+    margin-top: 1em;
+  }
+
+  & .toast__alert {
+    margin-left: -1px;
+    border-radius: .25em 0 0 .25em;
+    display: flex;
+    align-items: center;
+    justify-content: center;
+    width: 25px;
+    & path {
+      fill: white;
+    }
+    padding: .25em .5em;
+    ${(props) => getCSSForFlavor(props.flavor)}
+  }
+  & .toast__content {
+    padding: .5em 1em;
+    flex: 1;
+  }
+  &.toast--cancelled {
+    animation: immediatecancel .35s ease-out;
+  }
+  & .toast__cancel {
+    display: flex;
+    padding: .25em .5em;
+    & path {
+      fill: #54457F;
+    }
+    align-items: center;
+  }
+`;
+
+export const ToastAlert = ({ children, flavor, id }) => {
+  const { manuallyCancel, flavors } = React.useContext(ToastContext);
+  const internalFlavors = {
+    [flavors.success]: <FontAwesomeIcon icon={faCheckCircle} />,
+    [flavor.error]: <FontAwesomeIcon icon={faExclamationTriangle} />,
+  };
+  const alertIcon = internalFlavors[flavor];
+  const [cancel, setCancel] = useState(false);
+  const handleClick = () => {
+    setCancel(true);
+    manuallyCancel(id);
+  };
+  return (
+    <StyledToast
+      flavor={flavor}
+      ariaLive="polite"
+      className={cancel ? 'toast--cancelled' : ''}
+      onClick={handleClick}
+    >
+      <div className="toast__alert">
+        {alertIcon}
+      </div>
+      <div className="toast__content">
+        {children}
+      </div>
+      <div className="toast__cancel">
+        <FontAwesomeIcon icon={faTimes} />
+      </div>
+    </StyledToast>
+  );
+};
+
+ToastAlert.propTypes = {
+  children: PropTypes.node.isRequired,
+  flavor: PropTypes.string.isRequired,
+  id: PropTypes.number.isRequired,
+};
+
+export const ToastManager = ({ children, defaultTime }) => {
+  const toasts = useRef([]);
+  const [availableToastId, setAvailableToastId] = useState(0);
+  const [forceRerender, setForceRerender] = useState(0);
+
+  const contextData = {
```

Comment:
useMemo, you don't want to change the context reference nearly ever, as that would cause a remount of all your children, especially given this is one of the top level elements

Reply:
Hmm. I tried that but eslint is complaining that I'd need to pass `availableToastId` as a dependency of useMemo, which would cause it to rerender either way every time `toast()` is called.

Reply:
ESlint is doing it to generally warn against excluding values from the dependency list. You do have a potential issue as this stands, which is if two toasts are invoked in the same render you'll end up with two toasts of the same ID, I believe.

The exhaustive-deps is a general warn for "hey, this thing is a real bitch to debug if you don't get it right". We have it ignored in a ton of cases, but we require an explicit comment about why we're doing it

Reply:
Hmm yeah. 

>  which is if two toasts are invoked in the same render you'll end up with two toasts of the same ID, I believe

That ended up happening after all. So I'll have to pass on useMemo in this case :/

Reply:
You're misreading the outcome of that.

If you can't use useMemo here because of how you are generating IDs, how you are generating IDs is wrong.

You don't need it to be on state, it's just a number that needs to be unique, right? You can add that as a variable local to your JS file and get the same behavior w/o the use of useState which forces re-renders

Reply:
Setting the number to a new value is what causes the rerender that shows the alert though? Maybe we can review it together.